### ISSUE TITLE
Handling Multiple session with libnetconf2

### ISSUE THREAD

**Issue by vprasad** [CESNET/libnetconf2#121]
Hi Michal ,

I wanted to support multiple parallel netconf sessions so that multiple users can simultaneously do the login and do the transactions(get/get-config/edit-config,etc.) .

Currently i am spawning one worker thread that keeps on calling nc_accept in non blocking mode followed by nc_ps_poll.

However when i try to login from 2 separate window parallely ,the 2nd request gets hung.
So i wanted to ask whats the procedure to support multiple session .

As i can see in netopeer server code ,it spaws worker thread = NP2SRV_THREAD_COUNT +1 and calls accept and poll from each threads. continuously .Is this the procedure to support multiple sessions. 

Can you please brief me once ?
Also what can be max session that can be supported ?

Thanks,

**Comment by michalvasko** [comment]
Hi Vivek,
I do not quite understand what is it you are doing. You are asking about stuff implemented in `netopeer2-server` that you are implementing yourself, why not rather take the working server and modify it whatever way you need to?

Anyway, I think accepting 2 sessions in one thread should work simultaneously but it all depends on your exact implementation. Still, to achieve actual concurrency you always need to have more threads. There is no session number limitation in *libnetconf2*.

Regards,
Michal

**Comment by vprasad** [comment]
Hi Michal,

I can;t take directly the netopeer server code as it has dependencies on sysrepro,etc.
So i am trying to write the server code by referring to the netopeer server code.

For achieving multiple concurrent session i have figured out the following approach.Can you please review it once :-

1 .Call nc_accept in non -blocking mode from main thread .

2  .Once a session gets established and authenticated as a part of nc_accept, will add that to poll session struct via nc_ps_add_session and spawn a worker thread to do the nc_ps_poll with this session info.

3 .Worker thread will keep on polling on the particular session and will invoke the rpc's .Once session is closed/terminated ,will kill this thread and do the cleanup of removing session from poll struct.

Basically my approach would be to spawn a new worker thread for each session once the connection gets established as a part of nc_accept and do all the rpc as a part of new spawned worked thread.

By this approach my nc_accept will still be running as a part of main thread which would be continuing to accept any new connection..

Please review this approach once and suggest if this approach would work fine So that i can proceed for implementation.


Thanks,

**Comment by michalvasko** [comment]
Hi,
that is exactly what I meant by making your own modifications, rewriting sysrepo calls to actual datastore changes, whatever form it has in your case. I still think this would save you a lot of work.

Anyway, what you described should work. The whole difference in *netopeer2-server* is that we do not have a specific thread for accepting sessions or polling one session, just a generic worker that does everything. This way there is always the same (configurable) number of threads and none are created/terminated during server execution.

Regards,
Michal

**Comment by vprasad** [comment]
> Hi,
> that is exactly what I meant by making your own modifications, rewriting sysrepo calls to actual datastore changes, whatever form it has in your case. I still think this would save you a lot of work.
> 
> Anyway, what you described should work. The whole difference in _netopeer2-server_ is that we do not have a specific thread for accepting sessions or polling one session, just a generic worker that does everything. This way there is always the same (configurable) number of threads and none are created/terminated during server execution.
> 
> Regards,
> Michal

Hi Michal,

Thanks for info.

Just a query on the Netopeer server code with single worker thread .
So how will the following scenario  be handled :-

Suppose  user1 logged in and authenticated . However user1 is not invoking any rpc's.

At the same time another user2 is trying to login .But the netopeer server will be blocked in nc_ps_poll waiting for some rpc to happen for User1 .Till  nc_ps_poll time out happens for user1 or user1 does not sends any rpc ,user2 will not be able to login in this way if you have one worker thread doing the nc_accept as well as nc_ps_poll .
 
Also lets suppose 2 users ,user1 and user2 have logged in and authenticated .Now lets say  nc_ps_poll is waiting for some rpc to happen on user1 and at this moment whatever rpc User2 sends will not be accepted till User1 has give some rpc req as input or User1 nc_ps_poll timeout happens so that nc_ps_poll would proceed for user2.

These scenarios may be getting handle in Netopeer Server.Just wanted to understand how its getting handled.?

Thanks,

**Comment by vprasad** [comment]
> Hi,
> these scenarios are handled, naturally. If you look at `nc_accept()` and `nc_ps_poll()` calls you will see that `0` is used as timeout. That means none of the operations you mentioned block "until they happen" but immediately move on to another operation.
> 
> Regards,
> Michal

Thanks,
CLosing the issue.
